bgpv2: Adding pod cidr advertisement reconciler for BGPv2 APIs

harsimran-pabla opened the pull request:
Introducing pod cidr advertisement reconciler in BGP v2 reconcilers. 

- BGPv1 code change - adding option to set family field in path attributes. If no family is set, behavior is unchanged.
- Introducing new peer_advertisement cell inside reconcilers. This can be used by various advertisement reconcilers to get configured advertisement - per peer per family.
- Advertisement reconciler - generic layer to advertise/withdraw prefixes from BGP instance.
- PodCIDR reconciler - this uses all of the above to manage pod cidr advertisements.

```release-note
  bgpv2: Introducing pod cidr reconciler for bgpv2.
```

harsimran-pabla committed:
bgpv2: adding generic peer advertisement cell

Introducing generic peer advertisement cell which can be used by various
advertisement reconcilers to get configured advertisement option per
peer per address-family.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

rastislavs commented on pkg/bgpv1/gobgp/conversions.go:
```diff
@@ -96,6 +101,71 @@ func ToAgentPaths(paths []*gobgp.Path) ([]*types.Path, error) {
+	}
+}
+
+func toGoBGPAfi(a types.Afi) gobgp.Family_Afi {
```
Inverse functions to these (`toAgentAfi`, `toAgentSafi`) are located in `gobgp/state.go` ATM, maybe this would be a good opportunity to consolidate and move them into `conversions.go` as well?

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:
```diff
@@ -0,0 +1,138 @@
+	}
+}
+
+// GetAdvertisements can be called to get advertisements for given BGPAdvertisementType.
```
```suggestion
// GetAdvertisements can be called to get all configured advertisements of given BGPAdvertisementType for each peer.
```

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,153 @@
+
+	// an aset member which book keeps which universe it exists in
+	type member struct {
+		a     bool
```
What abut renaming `a` to `configured` and `b` to `advertised`? That would make the code a bit more readable.

rastislavs commented on pkg/bgpv1/types/state.go:
```diff
@@ -254,3 +258,10 @@ func ParseSafi(s string) Safi {
 	return ret
 }
+
+func ToAgentFamily(fam v2alpha1.CiliumBGPFamily) Family {
```
Maybe a good opportunity to rename this file to `conversions.go`, as it is (already) used not only for state data retrieval?

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+			var paths []*types.Path
+			for _, prefix := range desiredPrefixes {
+				path := types.NewPathForPrefix(prefix)
+				path.UpdateFamilyField(agentFamily)
```
What is the benefit of using the `UpdateFamilyField` helper instead of just `path.Family = agentFamily`?

harsimran-pabla replied:
I was thinking `path := types.NewPathForPrefix(prefix).PathWithFamily(agentFamily)`, somewhere in middle gave up on that idea. 

I'll go with `path.Family = agentFamily`, seems clear enough.

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+	return desiredFamilyAdverts, nil
+}
+
+func (r *ExportPodCIDRReconciler) reconcilerFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts ExportPodCIDRMetadata) error {
```
```suggestion
func (r *ExportPodCIDRReconciler) reconcileFamilyAdvertisements(ctx context.Context, p ReconcileParams, desiredFamilyAdverts ExportPodCIDRMetadata) error {
```

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+		if err != nil {
+			return err
+		}
+		runningFamilyAdverts[family] = runningAdverts
```
Could we do this in a single `ExportAdvertisementsReconciler` call per family?

harsimran-pabla replied:
I see there are two cases - delete, upsert. Both are called with different parameters of ExportAdvertisementReconciler; like enabled field. 

We could make a single call, but then deleted families might have stale family field with empty paths in metadata.
`runningFamilyAdverts[family] = nil`. I wanted to remove deleted family completely from runningFamilyAdverts map.

rastislavs replied:
ACK - understood, thx

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,153 @@
+		newAdverts []*types.Path
+	)
+
+	l.Debug("begin reconciling advertisements for BGP instance")
```
Isn't this a bit too verbose logging, given that we may call this multiple times per reconciler for each reconcile? Maybe we should log only when adding new advertisements / withdrawing paths?

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func (p *CiliumPeerAdvertisement) getFamilyAdvertisements(family v2alpha1.CiliumBGPFamilyWithAdverts, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) ([]v2alpha1.Advertisement, error) {
+	// get all advertisement CRDs
```
nit: all CRD objects, there is only one advertisement CRD

rastislavs commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:
```diff
@@ -0,0 +1,138 @@
+// We iterate over all advertisements ( available from local store ), select only those that match the advertisement
+// selector of the family.
+// Information of peer -> family -> advertisements is returned to the consumer.
+func (p *CiliumPeerAdvertisement) GetAdvertisements(conf *v2alpha1.CiliumBGPNodeInstance, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) (PeerAdvertisements, error) {
```
nit: maybe I would name this `GetConfiguredAdvertisements` or similar, to make it obvious that this is returning configuration, not state data.

rastislavs reviewed (changes requested):
Thanks, very nice work! Left some comments, but the overall approach looks good to me.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+		// logger for the reconciler
+		l = params.Logger.WithFields(logrus.Fields{
+			"reconciler": params.Reconciler,
+			"instance":   params.InstanceName,
```
Let's use logfield.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+}
+
+func (r *ExportPodCIDRReconciler) Name() string {
+	return "ExportPodCIDR"
```
Shall we take this opportunity to change the name to PodCIDR instead of ExportPodCIDR? I was feeling a bit weird about `Export` prefix because rest of the reconcilers also exports something, but not prefixed with `Export` (e.g., LBService).

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+// ExportPodCIDRMetadata is a map of advertisements per family, key is family type
+type ExportPodCIDRMetadata map[types.Family][]*types.Path
+
+func NewExportPodCIDRReconciler(params ExportPodCIDRReconcilerIn) ExportPodCIDRReconcilerOut {
```
Should we check the IPAM type here and not provide this reconciler for unsupported IPAM modes? We can also do it in the separated PR.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+		l.Debug("advertisements disabled for BGP instance")
+
+		for _, advrt := range params.CurrentAdvertisements {
+			l.WithField("NLRI", advrt.NLRI.String()).Debug("Withdrawing path")
```
Let's use log field

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go:
```diff
@@ -0,0 +1,254 @@
+		},
+	}
+
+	req := require.New(t)
```
Using this req inside the subtest below is invalid. You must create a new req within the t.Run(). 

Per documentation

```
FailNow must be called from the goroutine running the test
```

`require` package calls FailNow internally against given `t` and t.Run spawns a new goroutine to run function. Thus, we shouldn't call FailNow of parent req from subtest.

Same for other tests.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:
```diff
@@ -0,0 +1,138 @@
+// We iterate over all advertisements ( available from local store ), select only those that match the advertisement
+// selector of the family.
+// Information of peer -> family -> advertisements is returned to the consumer.
+func (p *CiliumPeerAdvertisement) GetConfiguredAdvertisements(conf *v2alpha1.CiliumBGPNodeInstance, selectAdvertTypes ...v2alpha1.BGPAdvertisementType) (PeerAdvertisements, error) {
```
So, the time complexity here is

`peers * families * advertisements` to search selected advertisements
`selectedAdvertisements * advertisement` to search target advertisement types
Then this happens for every advertisement reconciler (PodCIDR or PodIPPool, Service)

At this point we don't know the regular case, so might be a too early optimization, but I think we can pre-populate `advertisement type + family -> advertisements`. So that each reconciler can do an efficient query.

harsimran-pabla replied:
I think it is bit early to optimize. The main factor will be number of peers. Number of advertisements are pretty much constant.

 Linear scan [ Peers ] - O(n) ( number of peers )
 Linear scan [ Families ] - O(m) ( max 2 )
 Linear scan [ Advertisements ] - O(k) ( number of advertisements - 3-4 types, which is again filtered)

Overall complexity will be factor of number of peers. Even if we cache `type + families -> advertisements`, we will need to do linear scan per peer to get correct advertisement. Since number of families and advertisement configurations are expected to be small - I am not very concerned with this approach. 

We can def revisit this if we see it becoming a bottlenech - need some metrics :)

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+// ExportAdvertisementsReconciler reconciles the state of the BGP advertisements
+// with the provided toAdvertise list and returns a list of the advertisements
+// currently being announced.
+func ExportAdvertisementsReconciler(params *AdvertisementsReconcilerParams) ([]*types.Path, error) {
```
I'd name this `ReconcileAdvertisement` (verb).

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+type AdvertisementsReconcilerParams struct {
+	Logger     logrus.FieldLogger
+	Ctx        context.Context
+	Reconciler string
```
We don't need this parameter because the logger passed from reconcilers already contains the reconciler's name.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+	Enabled    bool
+
+	Instance     *instance.BGPInstance
+	InstanceName string
```
Here we already include instance name to the logfield: https://github.com/cilium/cilium/pull/30815/commits/cfd53394f83e379acb9f7344227eab28d4ff200f#diff-4f8b9238ec8c0592396726479067151675de57691dc7e50c09d181befe5055e9R137-R139

We should either remove this field and let reconcilers set the instance name or always rely on AdvertisementReconciler to set the logfield. I think the latter would be more generic and easier to use.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+	Logger     logrus.FieldLogger
+	Ctx        context.Context
+	Reconciler string
+	Enabled    bool
```
Can't we express this `Enabled` with zero length `ToAdvertise`?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+
+// getDesiredPathsPerFamily returns a map of desired paths per address family.
+// Note: This returns prefixes per address family. Global routing table will contain prefix per family not per neighbor.
+// Per neighbor advertisement will be controlled by BGP Policy.
```
Do we suppose to have this logic in this PR or are you planning to cover it in a separate PR?

harsimran-pabla replied:
It will be added in subsequent PR - one where we add policy. By default, we will then disable advertisements and only advertise based on the peer policy action.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go:
```diff
@@ -0,0 +1,254 @@
+		expectedAdverts    PeerAdvertisements
+	}{
+		{
+			name: "PodCIDR - expecting single advertisement for peer",
```
Are you planning to write tests for other reconciliation types? Otherwise, we can remove this `PodCIDR -` PeerAdvertisements should work for any advertisement types right?

harsimran-pabla replied:
Added more tests - to include v4/v6 families and other advertisement types.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements_test.go:
```diff
@@ -0,0 +1,254 @@
+		Spec: v2alpha1.CiliumBGPPeerConfigSpec{
+			Families: []v2alpha1.CiliumBGPFamilyWithAdverts{
+				{
+					CiliumBGPFamily: v2alpha1.CiliumBGPFamily{
+						Afi:  "ipv4",
+						Safi: "unicast",
+					},
```
I think we should test with dual stack. Currently, the case for multiple address families is not tested.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/peer_advertisements.go:
```diff
@@ -0,0 +1,138 @@
+		if err != nil {
+			return nil, err
+		}
+		if !exist {
+			lp.Debug("Peer config not found, skipping advertisement check")
+			continue
+		}
```
When we hit this case, `GetConfiguredAdvertisements` returns zero advertisements for this peer. It will remove all existing advertisements. We may have a situation that the resource store is not fully synchronized for PeerConfig, but we have BGPInstance. This is possible because the resource store is initialized asynchronously. In that case, we'll delete the route unexpectedly. I guess this will become a problem with GR at agent startup.

harsimran-pabla replied:
I think BGP startup should wait till stores are in sync. This is common problem in all reconcilers based on local store.   It is fixable imo - we need to wait at BGP CP initialization time for stores to get in sync.

YutaroHayakawa replied:
Ah, actually, GetByKey returns `ErrStoreUninitialized` error when the store is uninitialized. If we assume the store is already synchronized here, then when we hit the `ErrStoreUninitialized`, it's a bug. We should log it loudly. If we can catch that case in CI (maybe with Error log test?), that would be nice.

YutaroHayakawa replied:
The `!exist` case is expected because the user may refer to but forget to create the actual resource. Let's add a test case for this.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+				for _, existingPath := range existingPaths {
+					existingPathsSet.Insert(existingPath.NLRI.String())
+				}
+			}
```
Looking at the test coverage, this case has not been tested at all.

harsimran-pabla replied:
Added additional test case to have 2 peers. First peer will result in addition of the family. ExistingPathsSet will get populated by second peer.  In the test case, we will get 2 paths for each family ( 4 paths in total ).

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,194 @@
+				// skip if path already exists.
+				if existingPathsSet.Has(path.NLRI.String()) {
+					continue
+				}
```
We don't take path attribute or per-advertisement-type option (e.g., `spec.service`) changes into account here. How do we handle those cases?

harsimran-pabla replied:
Path attributes will be taken into consideration by policy reconciler.  Service reconciler will take into account additional options field. 

For Pod CIDR reconciler - I believe nothing changes.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+			}
+			continue
+		}
+		m.configured = true
```
Do we need this? At this point, member in the `aset` always have `configured: true` right?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/advertisements.go:
```diff
@@ -0,0 +1,151 @@
+			}
+			continue
+		}
+		m.advertised = true
```
Looking at the test coverage, we don't test this case at all. Means we don't test the case for `keep` case.

harsimran-pabla replied:
Modified test case to run the reconciler twice. In which case, we will run into keep case.

YutaroHayakawa reviewed (changes requested):
Thanks for working on this! Did an initial review.

YutaroHayakawa commented on pkg/bgpv1/gobgp/state.go:
```diff
@@ -141,100 +141,6 @@ func toAgentAfiSafiState(state *gobgp.AfiSafiState) *models.BgpPeerFamilies {
 	return res
 }
 
-// toAgentSessionState translates gobgp session state to cilium bgp session state.
```
Please split refactoring into separate commits.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,196 @@
+
+			// create existingPathsSet for easier existence check.
+			existingPathsSet := sets.New[string]()
+			existingPaths, ok := desiredFamilyAdverts[agentFamily]
```
Is !ok possible here?

harsimran-pabla replied:
For the first iteration of the peer, 'desiredFamilyAdverts` will be empty. Maybe there is confusion regarding `desiredPeerAdverts` and `desiredFamilyAdverts` ?

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+// PodCIDRMetadata is a map of advertisements per family, key is family type
+type PodCIDRMetadata map[types.Family][]*types.Path
```
Let's make this a struct containing this map. The current form is hard to extend.

And nit: PodCIDRMetadata => PodCIDRReconcilerMetadata

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,196 @@
+				paths = append(paths, path)
+			}
+
+			desiredFamilyAdverts[types.ToAgentFamily(family)] = append(desiredFamilyAdverts[types.ToAgentFamily(family)], paths...)
```
We already do agent family conversion here. Let's use this. 

https://github.com/cilium/cilium/pull/30815/commits/16b5a4fe58e59c6d566f3e996dabc1b7053d430f#diff-4f8b9238ec8c0592396726479067151675de57691dc7e50c09d181befe5055e9R114

harsimran-pabla replied:
good catch

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/pod_cidr.go:
```diff
@@ -0,0 +1,196 @@
+				ToAdvertise:           nil,
+			})
+			if err != nil {
+				return err
```
Reconcile advertisement may fail in the middle of the iteration. However, since we don't record that partial state in runningFamilyAdverts, it is possible to delete the same route twice. GoBGP returns an error in that case. Then we'll get into the failure loop here. I'd change ReconcileAdvertisement to return the current advertisement state even on error, so that we can reflect the actual state to the runningFamilyAdverts.

harsimran-pabla replied:
Modified Advertisement reconciler. It will return running advertisements in cases where there is an error.

I don't think we have a good way to test this. I will think more about it and add a test case in follow up PR.

YutaroHayakawa commented on pkg/bgpv1/manager/reconcilerv2/reconcilers.go:
```diff
@@ -32,8 +32,10 @@ type ConfigReconciler interface {
 }
 
 var ConfigReconcilers = cell.ProvidePrivate(
+	NewCiliumPeerAdvertisement,
```
Nit: CiliumPeerAdvertisement is not a ConfigReconciler. I'd provide it separately.

YutaroHayakawa commented on pkg/bgpv1/types/fake_router.go:
Why don't we use a real router?

harsimran-pabla replied:
I want to decouple gobgp from the reconciler tests. Couple of reasons
- Reconcilers should be agnostic of underlying router implementation.
- We run into port reuse issue, for eg if there is an issue in cleanup of gobgp instance. Rest of the tests will fail to start since port will already be in use.

YutaroHayakawa reviewed (changes requested):
Did a second review. Sorry for the delay 🙏

harsimran-pabla committed:
bgpv1: refactor conversion of family types

Add conversion of API CiliumBGPFamily to agent family type. Also,
refactored gobgp conversions.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: adding pod-cidr advertisement reconciler

Introducing pod-cidr advertisement reconciler as well as generic
advertisement layer which can be used by other advertisement reconcilers
to call lower layer BGP instance for advertising and withdrawing BGP
prefixes.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

harsimran-pabla committed:
bgpv2: adding pod-cidr advert tests

Adding unit tests for pod-cidr advertisement reconciler. Also,
introducing fake BGP router layer to decouple lower layer from unit
tests of reconcilers.

Signed-off-by: harsimran pabla <hpabla@isovalent.com>

YutaroHayakawa reviewed (approved):
Now looks good. Thanks!
